### PR TITLE
Add support for nullable property

### PR DESCRIPTION
**Pull request by joelwurtz**
Other way for #21 

WDYT @sagikazarmark ?

### CODE REVIEW

File: src/Generator/Normalizer/NormalizerGenerator.php [hunk 1]
```diff
@@ -72,20 +72,25 @@ protected function createSupportsNormalizationMethod($modelFqdn)
...
 
+            if ($property->isNullable()) {
+                $statements = array_merge($statements, $normalizationStatements);
+            }
+
             $statements[] = new Stmt\If_(
-                new Expr\BinaryOp\NotIdentical(new Expr\ConstFetch(new Name("null")), $propertyVar),
+                new Expr\BinaryOp\NotIdentical(new Expr\ConstFetch(new Name('null')), $propertyVar),
                 [
                     'stmts' => array_merge($normalizationStatements, [
```
Comment (sagikazarmark) [inline src/Generator/Normalizer/NormalizerGenerator.php]: What happens if it's nullable? Statements will be generated twice?

Comment (joelwurtz) [inline src/Generator/Normalizer/NormalizerGenerator.php]: Forgot the continue; keyword mybad

File: src/Generator/Normalizer/NormalizerGenerator.php [hunk 2]
```diff
@@ -72,20 +72,27 @@ protected function createSupportsNormalizationMethod($modelFqdn)
...
 
+        /** @var Property $property */
         foreach ($properties as $property) {
-            $propertyVar                                 = new Expr\MethodCall(new Expr\Variable('object'), $this->getNaming()->getPrefixedMethodName('get', $property->getName()));
+            $propertyVar = new Expr\MethodCall(new Expr\Variable('object'), $this->getNaming()->getPrefixedMethodName('get', $property->getName()));
             list($normalizationStatements, $outputVar) = $property->getType()->createNormalizationStatement($context, $propertyVar);
 
+            $normalizationStatements[] = new Expr\Assign(new Expr\PropertyFetch($dataVariable, sprintf("{'%s'}", $property->getName())), $outputVar);
+
+            if ($property->isNullable()) {
```
Comment (sagikazarmark) [inline src/Generator/Normalizer/NormalizerGenerator.php]: This way the data goes through all kinds of checks unnecessarily IMO. If we preserve the if statement and add an else then null can be set directly and all other checks (string, integer, etc) can be skipped.

**Comment by sagikazarmark** [comment]
How do you know if a multiple type is nullable?

File: src/Generator/Normalizer/NormalizerGenerator.php (hunk 2 above)
Comment (joelwurtz) [inline src/Generator/Normalizer/NormalizerGenerator.php]: Yeah it's unncessarily but it keeps thing cleaner in the architecture of generation (don't want to spread the logic of null everywhere), those kind of optimised path should be on another level (like with an ast optimizer which check null path and assignation)

**Comment by joelwurtz** [comment]
It's done in the ObjectGuesser

**Comment by sagikazarmark** [comment]
I don't see any change in denormalization. You use isset in denormalization, but that won't work with null values.

**Comment by joelwurtz** [comment]
Right, should be good now

**Comment by sagikazarmark** [comment]
Looks good to me. Although it lacks tests, doesn't it?

**Comment by joelwurtz** [comment]
Yes will add a better test system following this

**Comment by sagikazarmark** [comment]
Cool.

**Pull request merged**